persistence/sql/sql_process_store.go

```diff
@@ -159,6 +182,61 @@ func (p sqlProcessStoreImpl) doStartProcessTx(
 	}, err
 }
 
+func (p sqlProcessStoreImpl) doStopProcessTx(
+	ctx context.Context, tx extensions.SQLTransaction, request persistence.StopProcessRequest,
+) (*persistence.StopProcessResponse, error) {
+	curProcExecRow, err := p.session.SelectCurrentProcessExecution(ctx, request.Namespace, request.ProcessId)
+	if err != nil {
+		if p.session.IsNotFoundError(err) {
+			// early stop when there is no such process running
+			return &persistence.StopProcessResponse{
+				NotExists: true,
+			}, nil
+		}
+		return nil, err
+	}
+
+	// handle xdb_sys_process_executions
+	procExecRow, err := tx.SelectProcessExecutionForUpdate(ctx, curProcExecRow.ProcessExecutionId)
+	if err != nil {
+		return nil, err
+	}
+
+	procExecRow.IsCurrent = false
+
+	sequenceMaps, err := persistence.NewStateExecutionSequenceMapsFromBytes(procExecRow.StateExecutionSequenceMaps)
+	if err != nil {
+		return nil, err
+	}
+	sequenceMaps.PendingExecutionMap = map[string]map[int]bool{}
+	procExecRow.StateExecutionSequenceMaps, err = sequenceMaps.ToBytes()
+	if err != nil {
+		return nil, err
+	}
+
+	procExecRow.Status = persistence.ProcessExecutionStatusTerminated
+	if request.ProcessStopType == xdbapi.FAIL {
+		procExecRow.Status = persistence.ProcessExecutionStatusFailed
+	}
+
+	err = tx.UpdateProcessExecution(ctx, *procExecRow)
+	if err != nil {
+		return nil, err
+	}
+
+	// handle xdb_sys_async_state_executions
+	// find all related rows with the processExecutionId, and
+	// modify the wait_until/execute status from running to aborted
+	err = tx.UpdateAsyncStateExecutionToAbortRunning(ctx, curProcExecRow.ProcessExecutionId)
```

Comment:
Minor optimization idea: since we have loaded `PendingExecutionMap`, it would be better to check if there is any pending executions -- skip this update if there is no. 

It is possible that the process execution is still running when there is no state execution is running(because RPC can still trigger new state execution later)

Reply:
Good idea!